akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ShardCoordinator.scala

```diff
@@ -743,26 +774,35 @@ abstract class ShardCoordinator(
 
       case RebalanceDone(shard, ok) =>
         rebalanceWorkers -= sender()
-        if (ok)
-          log.debug("Rebalance shard [{}] completed successfully.", shard)
-        else
-          log.warning("Rebalance shard [{}] didn't complete within [{}].", shard, handOffTimeout.pretty)
-        // The shard could have been removed by ShardRegionTerminated
-        if (state.shards.contains(shard)) {
-          if (ok) {
+
+        val logContext =
+          if (gracefulShutdownInProgress(state.shards(shard))) "shutdown"
+          else "rebalance"
+
+        if (ok) {
+          log.debug("Shard [{}] {} completed successfully.", shard, logContext)
+
+          // The shard could have been removed by ShardRegionTerminated
+          if (state.shards.contains(shard)) {
             update(ShardHomeDeallocated(shard)) { evt =>
-              log.debug("Shard [{}] deallocated after rebalance", shard)
+              log.debug("Shard [{}] deallocated after {}", shard, logContext)
               state = state.updated(evt)
               clearRebalanceInProgress(shard)
               allocateShardHomesForRememberEntities()
               self.tell(GetShardHome(shard), ignoreRef)
             }
           } else {
-            // rebalance not completed, graceful shutdown will be retried
-            gracefulShutdownInProgress -= state.shards(shard)
             clearRebalanceInProgress(shard)
           }
+
         } else {
+          log.warning("Shard [{}] {} didn't complete within [{}].", shard, logContext, handOffTimeout.pretty)
+
+          // was that due to a graceful region shutdown?
+          // if so, consider the region as still alive and let it retry to gracefully shutdown later
+          if (gracefulShutdownInProgress(state.shards(shard)))
+            gracefulShutdownInProgress -= state.shards(shard)
+
```

Comment:
Extra attention to this part. 

The functionality didn't change, but I re-organised the if/else by reducing the amount of branching. 

In a nutshell, 

if `OK` and shard is in state, update the state, otherwise clear rebalance
if `NOK`, remove region from graceful shutdown if needed and clear rebalance

This is essentially what was happening before, but with more nesting and checks.